File: backend/app/services/order_service.py
from decimal import Decimal
from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import joinedload
from app.models.user import User
from app.models.product import Product, Cart
from app.models.order import Order, OrderItem
from app.models.delivery import Delivery
from app.schemas.order import CheckoutRequest
from app.repositories.order_repo import order_repo
from app.repositories.product_repo import product_repo
from app.repositories.delivery_repo import delivery_repo

class OrderService:
    async def place_order(self, db: AsyncSession, checkout_in: CheckoutRequest, current_user: User) -> Order:
        """Place an order atomically using SELECT FOR UPDATE pessimistic locking for stock safety under high load."""
        # 1. Fetch user's cart items
        cart_result = await db.execute(
            select(Cart)
            .filter(Cart.user_id == current_user.id)
            .options(joinedload(Cart.product))
        )
        cart_items = cart_result.scalars().all()
        
        if not cart_items:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Your cart is empty."
            )

        total_amount = Decimal("0.00")
        order_items_to_create = []

        try:
            # 2. Iterate and secure row locks on product items for inventory updates
            for item in cart_items:
                prod_result = await db.execute(
                    select(Product)
                    .filter(Product.id == item.product_id)
                    .with_for_update() # Pessimistic row locking
                )
                product = prod_result.scalars().first()

                if not product:
                    raise HTTPException(
                        status_code=status.HTTP_404_NOT_FOUND,
                        detail=f"Product with id {item.product_id} no longer exists."
                    )

                if product.stock < item.quantity:
                    raise HTTPException(
                        status_code=status.HTTP_400_BAD_REQUEST,
                        detail=f"Insufficient stock for product '{product.name}'. Available: {product.stock}, Requested: {item.quantity}"
                    )

                # Deduct inventory
                product.stock -= item.quantity
                item_total = product.price * item.quantity
                total_amount += item_total

                # Prepare OrderItem record
                order_item = OrderItem(
                    product_id=product.id,
                    quantity=item.quantity,
                    price=product.price
                )
                order_items_to_create.append(order_item)

            # 3. Create the parent Order record
            new_order = Order(
                user_id=current_user.id,
                total_amount=total_amount,
                status="Pending",
                items=order_items_to_create,
                shipping_address=checkout_in.shipping_address,
                customer_phone=checkout_in.phone,
                customer_name=checkout_in.full_name,
                payment_method=checkout_in.payment_method,
            )
            db.add(new_order)

            # 4. Clear shopping cart
            for item in cart_items:
                await db.delete(item)

            # 5. Automatically pre-generate the Delivery tracking record for the fulfillment queue
            import uuid
            tracking_number = f"TRK-{uuid.uuid4().hex[:12].upper()}"
            delivery_record = Delivery(
                order=new_order,
                status="assigned",
                tracking_number=tracking_number
            )
            db.add(delivery_record)

            # Commit transactional unit atomically
            await db.commit()
            await db.refresh(new_order)

            # Eager load relationships for clean API serialization
            final_result = await db.execute(
                select(Order)
                .filter(Order.id == new_order.id)
                .options(joinedload(Order.items).joinedload(OrderItem.product))
            )
            
            # Dispatch Async Background Jobs here if Celery is enabled
            # order_tasks.process_post_checkout.delay(new_order.id)
            
            return final_result.scalars().first()

        except Exception as e:
            await db.rollback()
            if isinstance(e, HTTPException):
                raise e
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"An error occurred during checkout processing: {str(e)}"
            )

order_service = OrderService()

File: backend/src/services/orderService.js
import { v4 as uuidv4 } from 'uuid';
import supabase from '../db/supabase.js';
import { AppError } from '../middleware/error.js';

export const orderService = {
  async placeOrder(checkoutData, currentUser) {
    const orderId = uuidv4();
    const deliveryId = uuidv4();
    const trackingNumber = `TRK-${uuidv4().replace(/-/g, '').substring(0, 12).toUpperCase()}`;

    // Call Postgres RPC for atomic checkout
    const { data: orderData, error: rpcError } = await supabase.rpc('place_order', {
      p_order_id: orderId,
      p_delivery_id: deliveryId,
      p_user_id: currentUser.id,
      p_shipping_address: checkoutData.shipping_address,
      p_phone: checkoutData.phone,
      p_full_name: checkoutData.full_name,
      p_payment_method: checkoutData.payment_method || 'card',
      p_tracking_number: trackingNumber
    });

    if (rpcError) {
      throw new AppError(400, rpcError.message);
    }

    // Eager load relationships for API serialization compatibility
    const { data: finalOrder, error: fetchError } = await supabase
      .from('orders')
      .select('*, items:order_items(*, product:products(*))')
      .eq('id', orderId)
      .single();

    if (fetchError || !finalOrder) {
      throw new AppError(500, `Error loading placed order details: ${fetchError?.message}`);
    }

    return finalOrder;
  },

  async cancelOrder(orderId, currentUser) {
    // 1. Fetch order with items
    const { data: order, error } = await supabase
      .from('orders')
      .select('*, items:order_items(*)')
      .eq('id', orderId)
      .single();

    if (error || !order) {
      throw new AppError(404, 'Order not found');
    }

    // 2. Validate ownership/role
    if (!currentUser.is_admin && order.user_id !== currentUser.id) {
      throw new AppError(403, 'You are not authorized to cancel this order');
    }

    // 3. Validate cancellable status
    const cancellableStatuses = ['Pending', 'Processing'];
    if (!cancellableStatuses.includes(order.status)) {
      throw new AppError(400, `Cannot cancel an order with status '${order.status}'. Only orders in ${cancellableStatuses.join(', ')} can be cancelled.`);
    }

    // 4. Restore product stocks
    for (const item of order.items) {
      if (item.product_id) {
        // Fetch current stock
        const { data: product } = await supabase
          .from('products')
          .select('stock')
          .eq('id', item.product_id)
          .single();
        
        if (product) {
          await supabase
            .from('products')
            .update({ stock: product.stock + item.quantity })
            .eq('id', item.product_id);
        }
      }
    }

    // 5. Update order status to Cancelled
    const { data: updatedOrder, error: updateError } = await supabase
      .from('orders')
      .update({ status: 'Cancelled' })
      .eq('id', orderId)
      .select('*, items:order_items(*, product:products(*))')
      .single();

    if (updateError) {
      throw new AppError(500, `Error updating order status: ${updateError.message}`);
    }

    return updatedOrder;
  }
};

export default orderService;
